ferrocene/doc/evaluation-report/src/self-test/tool-analysis.rst

```diff
@@ -0,0 +1,104 @@
+.. SPDX-License-Identifier: MIT OR Apache-2.0
+   SPDX-FileCopyrightText: The Ferrocene Developers
+
+.. default-domain:: qualification
+
+Tool Analysis
+=============
+
+Potential Errors
+----------------
+
+.. hazop-error::
+   :id: ERR_SELF_TEST_TARGET_CHECKS
+   :caused_by: USE_SELF_TEST_CHECK
+
+   | **Risk:** Invalid output
+   | **Detectable:** YES
+
+   Tool does not perform checks for some installed targets.
+
+.. hazop-error::
+   :id: ERR_SELF_TEST_FALSE_NEGATIVES
+   :caused_by: USE_SELF_TEST_CHECK
+
+   | **Risk:** Invalid output
+   | **Detectable:** NO
+
+   Tool fails to detect faulty toolchain installation.
+
+.. hazop-error::
+   :id: ERR_SELF_TEST_FALSE_POSITIVES
+   :caused_by: USE_SELF_TEST_CHECK
+
+   | **Risk:** Invalid output
+   | **Detectable:** YES
+
+   Tool falsely indicates a faulty toolchain installation.
+
+.. hazop-error::
+   :id: ERR_SELF_TEST_MODIFIES_TOOLCHAIN
+   :caused_by: USE_SELF_TEST_CHECK
+
+   | **Risk:** Invalid output
+   | **Detectable:** NO
+
+   Tool creates, changes, or deletes toolchain files/directories.
+
+.. hazop-error::
+   :id: ERR_SELF_TEST_VERSION_CHECK
+   :caused_by: USE_SELF_TEST_CHECK
+
+   | **Risk:** Invalid output
+   | **Detectable:** NO
+
+   Tool fails to detect that its toolchain is the wrong version.
+
+.. _self-test_avd:
+
+Detection Measures and Usage Restriction
+----------------------------------------
+
+Detection measures and usage restrictions are documented in the :doc:`safety
+manual <safety-manual:self-test/constraints>`.
+
+.. _ferrocene_self_test_iso_tool_classification:
+
+ISO 26262 Tool Classification
+-----------------------------
+
+During this analysis, we highlighted some of the potential errors concerning
+``ferrocene-self-test`` that impacts the safety-related software code.
+Hence, the tool impact is **TI2**.
+
+Moreover, this analysis shows us that the likelihood of detecting some of the
+potential errors is very low.
+Therefore, the tool error detection class is **TD3**.
+
+Using clause 11.4.5.4 in part 8 of the [|iso_ref|] standard, we can conclude
+that in the worst case the Tool Classification Level is **TCL3** up to ASIL D.
+Therefore, we choose the following qualification methods:
+
+- **1b.** Evaluation of the tool development process in accordance with 11.4.8
+- **1c.** Validation of the software tool in accordance with 11.4.9
+
+According to clause 11.4.2 in part 8 of the [|iso_ref|] standard,
+this choice depends on the user's software development life-cycle and their validation strategy.
+The user has the responsibility to determine whether this level, or a better one, is applicable.
+
+.. _ferrocene_self_test_iec_tool_classification:
+
+IEC 61508 Tool Classification
+-----------------------------
+
+``ferrocene-self-test`` verifies the installed toolchain,
+and is therefore an industrial [|iec_ref|] class T2 tool.
```

Comment:
I think it might rather be T3, because ferrocene-self-test could modify e.g. the rustc binary or a library which could result in wrong object code. See the description of the three classes below.

Quoting from [the Evaluation Plan](https://public-docs.ferrocene.dev/main/qualification/evaluation-plan/method.html#select-the-appropriate-qualification-method):

> - T1: the tool generates no outputs which can directly or indirectly contribute to the executable code (including data) of the safety-related system;
> - T2: the tool supports the test or verification of the design or executable code, where errors in the tool can fail to reveal defects but cannot directly create errors in the executable software;
> - T3: the tool generates outputs which can directly or indirectly contribute to the executable code of the safety-related system.

Reply:
I think we should add a step to ferrocene-self-test which ensures that no files were added, deleted or modified while ferrocene-self-test runs. Either by taking a snapshot of the ferrocene installation or a hash of the ferrocene installation or by checking the modification timestamps.

Reply:
the tool modifying the toolchain install would be a bug though, meaning that if it is T3, then all other tools are T3, since they all could have the bug

Reply:
True

Reply:
that check could also introduce the bug (and so on), so rather we have an integration that does that check

Reply:
this should rather be a test, to ensure that bug does not exist

Reply:
True

Reply:
3433c6fc6f1